# Introduction

Data is collected from 3 different sources:

twitter_archive_enhanced.csv = WeRateDogs twitter account archive, which is available in csv-format and will be downloaded manually.

image_predictions.tsv = Predictions what breed of dog is present in each tweet. This file will be downloaded programmatically by using the Requests library.

tweet_json.txt = Entire tweets from WeRateDogs Twitter archive. This file will be created  programmatically from WeRateDogs Twitter archive by using Twitter API  

# Gathering Data

In [39]:
import pandas as pd
import numpy as np

### Twitter Archive Data

In [148]:
df_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

## image-predictions.tsv 

In [3]:
from bs4 import BeautifulSoup
import requests 
import os

In [4]:
folder_name = 'project_wrangle_and_analyse_data'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [5]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

In [6]:
response = requests.get(url)

In [7]:
response

<Response [200]>

In [8]:
with open (os.path.join(folder_name, url.split('/')[-1]), mode = 'wb') as file:
    file.write(response.content)

In [9]:
df_predictions = pd.read_csv('project_wrangle_and_analyse_data/image-predictions.tsv', sep ='\t')

In [10]:
df_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

# tweet_json.txt 

In [ ]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret ='HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv

# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API

tweet_ids = df_archive.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)


In [11]:
import json
tweet_data = []

with open('tweet_json.txt','r') as json_file:  
    for line in json_file:
        json_data = json.loads(line)
        tweet_id = json_data['id']
        retweet_count = json_data['retweet_count']
        favorite_count = json_data['favorite_count']
        tweet_data.append({'retweet_count': int(retweet_count),
                        'favorite_count': int(favorite_count),
                        'tweet_id': int(tweet_id)})

df_tweets = pd.DataFrame(tweet_data, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

In [72]:
df_tweets

tweet_id  retweet_count  favorite_count
0     892420643555336193           8853           39467
1     892177421306343426           6514           33819
2     891815181378084864           4328           25461
3     891689557279858688           8964           42908
4     891327558926688256           9774           41048
...                  ...            ...             ...
2349  666049248165822465             41             111
2350  666044226329800704            147             311
2351  666033412701032449             47             128
2352  666029285002620928             48             132
2353  666020888022790149            532            2535

[2354 rows x 3 columns]

# Assessing Data 

### Twitter Archive Data

In [5]:
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [9]:
df_archive.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [11]:
df_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
309   835536468978302976                    NaN                  NaN   
113   870726314365509632           8.707262e+17           16487760.0   
1673  682303737705140231                    NaN                  NaN   
1976  672995267319328768                    NaN                  NaN   
1567  687826841265172480                    NaN                  NaN   

                      timestamp  \
309   2017-02-25 17:06:32 +0000   
113   2017-06-02 19:38:25 +0000   
1673  2015-12-30 20:54:22 +0000   
1976  2015-12-05 04:25:50 +0000   
1567  2016-01-15 02:41:12 +0000   

                                                 source  \
309   <a href="http://twitter.com/download/iphone" r...   
113   <a href="http://twitter.com/download/iphone" r...   
1673  <a href="http://twitter.com/download/iphone" r...   
1976  <a href="http://twitter.com/download/iphone" r...   
1567  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
309   RT @dog_rates: Meet Lola. Her hobbies include ...         8.352641e+17   
113   @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
1673  This is Todo. He's screaming because he doesn'...                  NaN   
1976  This is Pumpkin. He can look in two different ...                  NaN   
1567  This is Louis. He's a rollercoaster of emotion...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
309               4.196984e+09  2017-02-24 23:04:14 +0000   
113                        NaN                        NaN   
1673                       NaN                        NaN   
1976                       NaN                        NaN   
1567                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
309   https://www.gofundme.com/lolas-life-saving-sur...                12   
113                                                 NaN                10   
1673  https://twitter.com/dog_rates/status/682303737...                 9   
1976  https://twitter.com/dog_rates/status/672995267...                 8   
1567  https://twitter.com/dog_rates/status/687826841...                12   

      rating_denominator     name doggo floofer pupper puppo  
309                   10     Lola  None    None   None  None  
113                   10     None  None    None   None  None  
1673                  10     Todo  None    None   None  None  
1976                  10  Pumpkin  None    None   None  None  
1567                  10    Louis  None    None   None  None

In [6]:
df_archive.shape

(2356, 17)

In [8]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [81]:
df_archive["retweeted_status_id"] 

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
2351   NaN
2352   NaN
2353   NaN
2354   NaN
2355   NaN
Name: retweeted_status_id, Length: 2356, dtype: float64

In [12]:
# Let's check if there are any row duplicates 
df_archive.duplicated().sum()

0

In [16]:
df_archive["tweet_id"].nunique()

2356

In [17]:
df_archive["source"].value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [30]:
df_archive["name"]

0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
          ...   
2351        None
2352           a
2353           a
2354           a
2355        None
Name: name, Length: 2356, dtype: object

In [51]:
df_archive["name"].value_counts()

None       745
a           55
Charlie     12
Oliver      11
Cooper      11
          ... 
Shelby       1
Gilbert      1
Laika        1
Deacon       1
Dug          1
Name: name, Length: 957, dtype: int64

In [163]:
df_archive_clean['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

### Observations: 

 
Data Types

- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id and retweeted_status_user_id are FLOATS  
- timestamp and retweeted_status_timestamp are OBJECTS

Missing values

- in_reply_to_status_id and in_reply_to_user_id only 78 non null values
- retweeted_status_id, retweeted_status_user_id and retweeted_status_timestamp only 181 non null values
- expanded_urls 2297 non null values
- Missig values are stated as None instead of NaN

Other potential issues

- It looks like that there are wrong dog names, for example: a, unacceptable.

- Source column has html code 

- According to wikipedia (link below) scale of the rating_denominator should be 10. Some of the values in rating_denominator column are off the chart. 

https://en.wikipedia.org/wiki/WeRateDogs

### Image Predictions

In [66]:
df_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [177]:
df_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [67]:
df_predictions["p1"].value_counts()

Labrador_retriever    104
golden_retriever       92
Cardigan               73
Chihuahua              44
Pomeranian             42
                     ... 
white_wolf              1
barracouta              1
spatula                 1
turnstile               1
tree_frog               1
Name: p2, Length: 405, dtype: int64

In [179]:
df_predictions['p1'].str.islower().sum()

1135

In [68]:
df_predictions["p2"].value_counts()

Labrador_retriever    104
golden_retriever       92
Cardigan               73
Chihuahua              44
Pomeranian             42
                     ... 
white_wolf              1
barracouta              1
spatula                 1
turnstile               1
tree_frog               1
Name: p2, Length: 405, dtype: int64

In [180]:
df_predictions['p2'].str.islower().sum()

1096

In [69]:
df_predictions["p3"].value_counts()

Labrador_retriever    79
Chihuahua             58
golden_retriever      48
Eskimo_dog            38
kelpie                35
                      ..
lampshade              1
goldfish               1
mountain_tent          1
tiger_cat              1
wallet                 1
Name: p3, Length: 408, dtype: int64

### Observation:

The titles of the columns are not very clear and easy to understand  

Some of the dog breeds in p1, p2 and p3 columns are sometimes capitalised and sometimes not.

### Tweets 

In [74]:
df_tweets.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [75]:
df_tweets.tail()

tweet_id  retweet_count  favorite_count
2349  666049248165822465             41             111
2350  666044226329800704            147             311
2351  666033412701032449             47             128
2352  666029285002620928             48             132
2353  666020888022790149            532            2535

In [76]:
df_tweets.sample(5)

tweet_id  retweet_count  favorite_count
239   846514051647705089          13076           48410
402   824297048279236611           4463           16625
2157  669571471778410496           1085            1684
432   820690176645140481           3716           13518
73    878316110768087041           6965               0

In [77]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


# Cleaning Data

In [149]:
# Copies of each DataFrame
df_archive_clean = df_archive.copy()
df_predictions_clean = df_predictions.copy()
df_tweets_clean = df_tweets.copy()

### Twitter Archive Data

In [182]:
df_archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp              source  \
0  2017-08-01 16:23:56 +0000  Twitter for iPhone   
1  2017-08-01 00:17:27 +0000  Twitter for iPhone   
2  2017-07-31 00:18:03 +0000  Twitter for iPhone   
3  2017-07-30 15:58:51 +0000  Twitter for iPhone   
4  2017-07-29 16:00:24 +0000  Twitter for iPhone   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### q1 - Removing useless columns

In [222]:
df_archive_clean = df_archive_clean.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'in_reply_to_status_id',
'in_reply_to_user_id', 'in_reply_to_user_id'],axis = 1)

In [225]:
df_archive_clean.head(2)

tweet_id                  timestamp              source  \
0  892420643555336193  2017-08-01 16:23:56 +0000  Twitter for iPhone   
1  892177421306343426  2017-08-01 00:17:27 +0000  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [251]:
df_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2356 non-null   object
 1   timestamp           2356 non-null   object
 2   source              2356 non-null   object
 3   text                2356 non-null   object
 4   expanded_urls       2297 non-null   object
 5   rating_numerator    2356 non-null   int64 
 6   rating_denominator  2356 non-null   int64 
 7   name                2247 non-null   object
 8   doggo               2356 non-null   object
 9   floofer             2356 non-null   object
 10  pupper              2356 non-null   object
 11  puppo               2356 non-null   object
dtypes: int64(2), object(10)
memory usage: 221.0+ KB


#### q2 - Change data type from object to timestamp 

In [125]:
 df_archive_clean['timestamp'] = pd.to_datetime(df_archive_clean.timestamp)

In [126]:
#Let's test the change
df_archive_clean['timestamp'].dtype

datetime64[ns, UTC]

#### q3 - Replace wrong dog names to NaN

In [172]:
#Let's check those wrong names in more detailed level
wrong_dog_names = []

for name in df_archive_clean['name']:
    if name[0].islower():
        wrong_dog_names.append(name)
        print(name, end =', ')

such, a, quite, quite, quite, not, one, incredibly, a, mad, an, very, a, very, just, my, one, not, his, one, a, a, a, an, very, actually, a, just, getting, mad, very, this, unacceptable, all, a, old, a, infuriating, a, a, a, an, a, a, very, getting, just, a, the, the, actually, by, a, officially, a, the, the, a, a, a, a, life, a, one, a, a, a, light, just, space, a, the, a, a, a, a, a, a, a, a, a, an, a, the, a, a, a, a, a, a, a, a, a, a, a, quite, a, an, a, an, the, the, a, a, an, a, a, a, a, 

In [173]:
change_names = df_archive_clean[df_archive_clean['name'].isin(wrong_dog_names)] 

In [174]:
for names in change_names.index:
    df_archive_clean.loc[names, 'name'] = np.nan

#### q4 - Replace 'None' to NaN

In [138]:
 
df_archive_clean['name'] = df_archive_clean['name'].replace('None', np.nan)

Let's test both wrong names (dog names and 'None') changes

In [139]:
df_archive_clean['name']

0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
          ...   
2351         NaN
2352         NaN
2353         NaN
2354         NaN
2355         NaN
Name: name, Length: 2356, dtype: object

In [175]:
df_archive_clean['name'].value_counts()

None       745
Charlie     12
Oliver      11
Lucy        11
Cooper      11
          ... 
Shelby       1
Aubie        1
Laika        1
Deacon       1
Dug          1
Name: name, Length: 932, dtype: int64

In [147]:
df_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2356 non-null   int64              
 1   in_reply_to_status_id       78 non-null     float64            
 2   in_reply_to_user_id         78 non-null     float64            
 3   timestamp                   2356 non-null   datetime64[ns, UTC]
 4   source                      0 non-null      object             
 5   text                        2356 non-null   object             
 6   retweeted_status_id         181 non-null    float64            
 7   retweeted_status_user_id    181 non-null    float64            
 8   retweeted_status_timestamp  181 non-null    object             
 9   expanded_urls               2297 non-null   object             
 10  rating_numerator            2356 non-null   int64           

####  q5 - Remove html code from source column.
 

In [150]:
df_archive_clean['source'] 

0       <a href="http://twitter.com/download/iphone" r...
1       <a href="http://twitter.com/download/iphone" r...
2       <a href="http://twitter.com/download/iphone" r...
3       <a href="http://twitter.com/download/iphone" r...
4       <a href="http://twitter.com/download/iphone" r...
                              ...                        
2351    <a href="http://twitter.com/download/iphone" r...
2352    <a href="http://twitter.com/download/iphone" r...
2353    <a href="http://twitter.com/download/iphone" r...
2354    <a href="http://twitter.com/download/iphone" r...
2355    <a href="http://twitter.com/download/iphone" r...
Name: source, Length: 2356, dtype: object

In [151]:
df_archive_clean['source'] = df_archive_clean['source'].str.extract('^<a.+>(.+)</a>$')

In [152]:
df_archive_clean["source"] 

0       Twitter for iPhone
1       Twitter for iPhone
2       Twitter for iPhone
3       Twitter for iPhone
4       Twitter for iPhone
               ...        
2351    Twitter for iPhone
2352    Twitter for iPhone
2353    Twitter for iPhone
2354    Twitter for iPhone
2355    Twitter for iPhone
Name: source, Length: 2356, dtype: object

In [153]:
df_archive_clean["source"].value_counts()

Twitter for iPhone     2221
Vine - Make a Scene      91
Twitter Web Client       33
TweetDeck                11
Name: source, dtype: int64

#### q6 - Change all the values in rating_denominator column were the value is not 10

In [160]:
df_archive_clean['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [231]:
not_10 = df_archive_clean[df_archive_clean['rating_denominator'] != 10]

In [232]:
not_10['tweet_id']

313     835246439529840640
342     832088576586297345
433     820690176645140481
516     810984652412424192
784     775096608509886464
902     758467244762497024
1068    740373189193256964
1120    731156023742988288
1165    722974582966214656
1202    716439118184652801
1228    713900603437621249
1254    710658690886586372
1274    709198395643068416
1351    704054845121142784
1433    697463031882764288
1598    686035780142297088
1634    684225744407494656
1635    684222868335505415
1662    682962037429899265
1663    682808988178739200
1779    677716515794329600
1843    675853064436391936
2335    666287406224695296
Name: tweet_id, dtype: object

In [240]:
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 835246439529840640, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 832088576586297345, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 820690176645140481, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 810984652412424192, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 775096608509886464, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 758467244762497024, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 740373189193256964, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 731156023742988288, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 722974582966214656, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 716439118184652801, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 713900603437621249, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 710658690886586372, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 709198395643068416, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 704054845121142784, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 697463031882764288, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 686035780142297088, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 684225744407494656, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 684222868335505415, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 682962037429899265, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 682808988178739200, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 677716515794329600, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 675853064436391936, ['rating_denominator']] = 10
df_archive_clean['rating_denominator'] = df_archive_clean.loc[df_archive_clean['tweet_id'] == 666287406224695296, ['rating_denominator']] = 10
 

In [241]:
df_archive_clean['rating_denominator'].value_counts()

10    2356
Name: rating_denominator, dtype: int64

### Image Predictions 

In [184]:
df_predictions_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#### q7 - Change the column names to improve understandability  

In [213]:
# change the column names
df_predictions_clean = df_predictions_clean.rename(columns = {'p1':'first_prediction', 'p1_conf': 'first_confidence', 'p1_dog': 'first_dog',
                                  'p2': 'second_prediction', 'p2_conf': 'second_confidence', 'p2_dog': 'second_dog',
                                  'p3': 'third_prediction', 'p3_conf': 'third_confidence', 'p3_dog': 'third_dog'} )

In [214]:
df_predictions_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num        first_prediction  first_confidence  first_dog  \
0           1  Welsh_springer_spaniel          0.465074       True   
1           1                 redbone          0.506826       True   
2           1         German_shepherd          0.596461       True   
3           1     Rhodesian_ridgeback          0.408143       True   
4           1      miniature_pinscher          0.560311       True   
...       ...                     ...               ...        ...   
2070        2                  basset          0.555712       True   
2071        1             paper_towel          0.170278      False   
2072        1               Chihuahua          0.716012       True   
2073        1               Chihuahua          0.323581       True   
2074        1                  orange          0.097049      False   

       second_prediction  second_confidence  second_dog  \
0                 collie           0.156665        True   
1     miniature_pinscher           0.074192        True   
2               malinois           0.138584        True   
3                redbone           0.360687        True   
4             Rottweiler           0.243682        True   
...                  ...                ...         ...   
2070    English_springer           0.225770        True   
2071  Labrador_retriever           0.168086        True   
2072            malamute           0.078253        True   
2073            Pekinese           0.090647        True   
2074               bagel           0.085851       False   

                 third_prediction  third_confidence  third_dog  
0               Shetland_sheepdog          0.061428       True  
1             Rhodesian_ridgeback          0.072010       True  
2                      bloodhound          0.116197       True  
3              miniature_pinscher          0.222752       True  
4                        Doberman          0.154629       True  
...                           ...               ...        ...  
2070  German_short-haired_pointer          0.175219       True  
2071                      spatula          0.040836      False  
2072                       kelpie          0.031379       True  
2073                     papillon          0.068957       True  
2074                       banana          0.076110      False  

[2075 rows x 12 columns]

#### q8 - Capitalize dog breed names in columns firts_prediction, second_prediction and third_prediction columns

In [215]:
df_predictions_clean['first_prediction'] = df_predictions_clean['first_prediction'].str.capitalize()

In [216]:
df_predictions_clean['second_prediction'] = df_predictions_clean['second_prediction'].str.capitalize()

In [217]:
df_predictions_clean['third_prediction'] = df_predictions_clean['third_prediction'].str.capitalize()

In [200]:
#Let's test how it works
df_predictions_clean['first_prediction'].str.islower().sum()

0

In [218]:
#Let's test how it works
df_predictions_clean['second_prediction'].str.islower().sum()

0

In [219]:
#Let's test how it works
df_predictions_clean['third_prediction'].str.islower().sum()

0

In [220]:
df_predictions_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num        first_prediction  first_confidence  first_dog  \
0           1  Welsh_springer_spaniel          0.465074       True   
1           1                 Redbone          0.506826       True   
2           1         German_shepherd          0.596461       True   
3           1     Rhodesian_ridgeback          0.408143       True   
4           1      Miniature_pinscher          0.560311       True   
...       ...                     ...               ...        ...   
2070        2                  Basset          0.555712       True   
2071        1             Paper_towel          0.170278      False   
2072        1               Chihuahua          0.716012       True   
2073        1               Chihuahua          0.323581       True   
2074        1                  Orange          0.097049      False   

       second_prediction  second_confidence  second_dog  \
0                 Collie           0.156665        True   
1     Miniature_pinscher           0.074192        True   
2               Malinois           0.138584        True   
3                Redbone           0.360687        True   
4             Rottweiler           0.243682        True   
...                  ...                ...         ...   
2070    English_springer           0.225770        True   
2071  Labrador_retriever           0.168086        True   
2072            Malamute           0.078253        True   
2073            Pekinese           0.090647        True   
2074               Bagel           0.085851       False   

                 third_prediction  third_confidence  third_dog  
0               Shetland_sheepdog          0.061428       True  
1             Rhodesian_ridgeback          0.072010       True  
2                      Bloodhound          0.116197       True  
3              Miniature_pinscher          0.222752       True  
4                        Doberman          0.154629       True  
...                           ...               ...        ...  
2070  German_short-haired_pointer          0.175219       True  
2071                      Spatula          0.040836      False  
2072                       Kelpie          0.031379       True  
2073                     Papillon          0.068957       True  
2074                       Banana          0.076110      False  

[2075 rows x 12 columns]

In [ ]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='left')

In [ ]:
inner_join_df= pd.merge(df1, df2, on ='Customer_id', how='inner')

In [246]:
df_archive_clean = pd.merge(df_archive_clean, df_tweets_clean, on='tweet_id', how = 'left')

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

In [249]:
df_archive_clean = df_archive_clean.join(df_tweets_clean, how='left', on='tweet_id')

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

In [ ]:
xz<<x<x<